`say` type incorrectly inferred as `never` when using `SlackEventMiddlewareArgs<'pin_added'>`

### Description

The `SlackEventMiddlewareArgs` type contains a `say` property that is set to type `SayFn` if the event contains a `channel` property or a `item.channel` property. The `PinAddedEvent` type _does_ have an `item.channel` property, however the `WhenEventHasChannelContext` returns `never` nonetheless. 

For example, the following code causes the TS error 

> `error TS2349: This expression is not callable. Type 'never' has no call signatures.` 

but while running the code, you can clearly see that `say` correctly exists as a `function`:

```
async function handlePinAdded(
  { say }
    : SlackEventMiddlewareArgs<'pin_added'>): Promise<void> {
  await say("message");
}
```

I noticed that the `PinAddedEvent.item` property is a `Record<string, unknown>`, so it may be that the `WhenEventHasChannelContext` type function always returns `never` due to the `unknown`, but I'm not entirely clear on how that field populates. The `ReactionAddedEvent` also has this problem, likely because its `item` property is a union type of `ReactionMessageItem | ReactionFileItem | ReactionFileCommentItem` and only `ReactionMessageItem` has a `channel` field. 

As a positive case, using `SlackEventMiddlewareArgs<'message'>` works correctly and TS infers `say` to be a `SayFn`.


### What type of issue is this? (place an `x` in one of the `[ ]`)
- [x] bug
- [ ] enhancement (feature request)
- [ ] question
- [ ] documentation related
- [ ] example code related
- [ ] testing related
- [ ] discussion

### Requirements (place an `x` in each of the `[ ]`)
* [x] I've read and understood the [Contributing guidelines](https://github.com/slackapi/bolt/blob/main/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).
* [x] I've searched for any related issues and avoided creating a duplicate issue.

---

### Bug Report

Filling out the following details about bugs will help us solve your issue sooner.

#### Reproducible in:

package version: 3.11.x

node version: 14.x

OS version(s): macOS Big Sur 11.6.6

#### Steps to reproduce:

1. Register an event handler on your app (for example, ` app.event('pin_added', handlePinAdded);`
2. Inside the function, attempt to infer the type of the `say` property on the event.
3. Call `say('some message);'` inside the event handler.

#### Expected result:

TS correctly infers the type of `say` to be `SayFn` and transpiles without errors.

#### Actual result:

TS incorrectly infers the type of `say` to be `never`.

#### Attachments:
![Screen Shot 2022-06-02 at 3 58 01 PM](https://user-images.githubusercontent.com/6301599/171737123-4aec5d68-fde2-4496-8c94-13692df292c6.png)
![Screen Shot 2022-06-02 at 3 58 18 PM](https://user-images.githubusercontent.com/6301599/171737145-3dc79c6e-24f7-4af9-8377-3b5460f8bcd7.png)

--- a/src/types/events/base-events.ts
+++ b/src/types/events/base-events.ts
@@ -585,12 +585,38 @@ export interface MemberLeftChannelEvent {
 
 export type MessageEvent = AllMessageEvents;
 
+export interface PinnedMessageItem {
+  client_msg_id?: string;
+  type: string;
+  app_id?: string;
+  team?: string;
+  user: string;
+  bot_id?: string;
+  bot_profile?: BotProfile;
+  text?: string;
+  attachments?: MessageAttachment[];
+  blocks?: (KnownBlock | Block)[];
+  pinned_to?: string[];
+  permalink: string;
+}
+export interface PinnedFileItem {
+  id: string;
+  // TODO: Add all other possible properties here
+}
+
+export interface PinnedItem {
+  type: string;
+  channel: string;
+  created_by: string;
+  created: number;
+  message?: PinnedMessageItem;
+  file?: PinnedFileItem;
+}
 export interface PinAddedEvent {
   type: 'pin_added';
   user: string;
   channel_id: string;
-  // TODO: incomplete, should be message | file | file comment (deprecated)
-  item: Record<string, unknown>;
+  item: PinnedItem;
   item_user: string;
   pin_count: string;
   pinned_info: {
@@ -600,13 +626,11 @@ export interface PinAddedEvent {
   };
   event_ts: string;
 }
-
 export interface PinRemovedEvent {
   type: 'pin_removed';
   user: string;
   channel_id: string;
-  // TODO: incomplete, should be message | file | file comment (deprecated)
-  item: Record<string, unknown>;
+  item: PinnedItem;
   item_user: string;
   pin_count: string;
   pinned_info: {

--- a/types-tests/event.test-d.ts
+++ b/types-tests/event.test-d.ts
@@ -1,5 +1,5 @@
 import { expectNotType, expectType } from 'tsd';
-import { App, SlackEvent, AppMentionEvent, ReactionAddedEvent, ReactionRemovedEvent, UserHuddleChangedEvent, UserProfileChangedEvent, UserStatusChangedEvent } from '..';
+import { App, SlackEvent, AppMentionEvent, ReactionAddedEvent, ReactionRemovedEvent, UserHuddleChangedEvent, UserProfileChangedEvent, UserStatusChangedEvent, PinAddedEvent, SayFn, PinRemovedEvent } from '..';
 
 const app = new App({ token: 'TOKEN', signingSecret: 'Signing Secret' });
 
@@ -47,6 +47,22 @@ expectType<void>(
   app.event('user_status_changed', async ({ event }) => {
     expectType<UserStatusChangedEvent>(event);
     expectNotType<SlackEvent>(event);
+    await Promise.resolve(event);
+  })
+);
+
+expectType<void>(
+  app.event('pin_added', async ({ say, event }) => {
+    expectType<SayFn>(say);
+    expectType<PinAddedEvent>(event);
+    await Promise.resolve(event);
+  })
+);
+
+expectType<void>(
+  app.event('pin_removed', async ({ say, event }) => {
+    expectType<SayFn>(say);
+    expectType<PinRemovedEvent>(event);
     await Promise.resolve(event);
   })
 );

